Bug 1780806 - Allow the uploader to signal it is done handling uploads.

badboy opened the pull request:
Putting this up so it's up.
I want to discuss this with chutten when he's back.

I'm not super happy with it because the `UploadResult::Done` needs to be handled on a different layer than all the other `UploadResult` variants.
(It's also only implemented for the RLB right now)

chutten commented on glean-core/rlb/src/test.rs:
```diff
@@ -1131,3 +1135,73 @@ fn test_boolean_get_num_errors() {
+
+    let map = call_count.lock().unwrap();
+    assert_eq!(2, map.len(), "should have launched 2 uploader threads");
+    for (_tid, &count) in &*map {
```
How many threads do we use to call out to the uploader, anyway? If it's one per submission, then this won't properly ensure we've handled `Done` properly

chutten commented on glean-core/rlb/src/net/mod.rs:
```diff
@@ -97,21 +98,35 @@ impl UploadManager {
                                 request.headers.into_iter().collect();
-                            let result = inner.uploader.upload(upload_url, request.body, headers);
+                            let result =
+                                match inner.uploader.upload(upload_url, request.body, headers) {
```
If you're concerned about maintaining the abstraction, you could drill through a return value from `PingUploadManager::process_ping_upload_response`. Then you could add this `UploadResult::Done` case to its `match`, returning `PingUploadTask::Done` or something. Then handle the correct `Done` here.

badboy replied:
I'm not sure I understand what you mean.
Right now `glean_process_ping_upload_response` doesn't return anything. Are you suggesting it _should_ return something that this uploading loop handles to determine the next step?
(That again doesn't seem _too_ bad I think)

chutten replied:
That's precisely what I meant. Have the thing that handles `UploadResult::*` handle this new one, then have it return the result all the way back out here.

badboy commented:
Ok, I reworked this a bit now.
The `Done` is passed through and then `glean_process_ping_upload_response` returns something that instructs the upload loop whether to continue or end.  I also implemented that for Kotlin now to show that it works.

Additionally I reworked the test. Instead of arbitrary sleep times it now uses a barrier to sync up the 2 threads. That should be much more reliable.

chutten commented on glean-core/android/src/main/java/mozilla/telemetry/glean/scheduler/PingUploadWorker.kt:
```diff
@@ -112,13 +113,19 @@ class PingUploadWorker(context: Context, params: WorkerParameters) : Worker(cont
+                    when (action) {
+                        UploadTaskAction.CONTINUE -> continue
+                        UploadTaskAction.END -> break
+                    }
```
No longer RLB-only, I see.

chutten commented on glean-core/rlb/src/net/mod.rs:
```diff
@@ -99,19 +100,25 @@ impl UploadManager {
+                            }
                         }
                         PingUploadTask::Wait { time } => {
+                            log::trace!("Instructed to wait for {:?}ms", time);
```
If we're tracing the other `PingUploadTask::*` we should trace `PingUploadTask::Upload`, no?

chutten commented on .detekt.yml:
```diff
@@ -9,6 +9,14 @@ style:
     active: false
+
+  # Go ahead and jump (Jump!)
+  # Go ahead and jump
```
👏

chutten commented on glean-core/rlb/src/test.rs:
```diff
@@ -1131,3 +1134,80 @@ fn test_boolean_get_num_errors() {
+    custom_ping.submit(None);
+
+    // Sync up with the upload thread.
+    barrier.wait();
```
Ah, and if the threading model changes (ie, if each `submit` starts getting its own thread) this test will fail by locking itself?

badboy replied:
yep

badboy replied:
well, in theory. *goes back to the code* why doesn't this fail now? oO

badboy replied:
ugh, turns out I'm still running into timing issues here. Let me once again rethink the test.

chutten commented on glean-core/src/upload/result.rs:
```diff
@@ -32,6 +32,14 @@ pub enum UploadResult {
     },
+
+    /// Signal that this uploader is done with work
+    /// and doesn't accept new work.
```
```suggestion
    /// and won't accept new work.
```

chutten reviewed (commented):
Needs CHANGELOG
